java/ql/src/experimental/Security/CWE/CWE-730/RegexInjection.ql

```diff
@@ -0,0 +1,101 @@
+/**
+ * @name Regular expression injection
+ * @description User input should not be used in regular expressions without first being sanitized,
+ *              otherwise a malicious user may be able to provide a regex that could require
+ *              exponential time on certain inputs.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id java/regex-injection
+ * @tags security
+ *       external/cwe/cwe-730
+ *       external/cwe/cwe-400
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * A data flow sink for untrusted user input used to construct regular expressions.
+ */
+class RegexSink extends DataFlow::ExprNode {
+  RegexSink() {
+    exists(MethodAccess ma, Method m | m = ma.getMethod() |
+      (
+        m.getDeclaringType().hasQualifiedName("java.lang", "String") and
+        (
```

Comment:
Here and below you could probably get rid of these parentheses since the formulas are only combined with `and` which has a higher precedence than `or` (formatting will also group this accordingly).

Reply:
No, it is already auto-formatted. All parentheses are required.

Reply:
I feel rather certain that you can omit them due to precedence rules, the following should be equivalent to your current code (already formatted):
```ql
exists(MethodAccess ma, Method m | m = ma.getMethod() |
  m.getDeclaringType() instanceof TypeString and
  ma.getArgument(0) = this.asExpr() and
  m.hasName(["matches", "split", "replaceFirst", "replaceAll"])
  or
  m.getDeclaringType().hasQualifiedName("java.util.regex", "Pattern") and
  ma.getArgument(0) = this.asExpr() and
  m.hasName(["compile", "matches"])
  or
  m.getDeclaringType().hasQualifiedName("org.apache.commons.lang3", "RegExUtils") and
  ma.getArgument(1) = this.asExpr() and
  m.getParameterType(1).(Class) instanceof TypeString and
  m.hasName([
      "removeAll", "removeFirst", "removePattern", "replaceAll", "replaceFirst",
      "replacePattern"
    ])
)
```